src/master-playlist-controller.js

```diff
@@ -0,0 +1,674 @@
+    externHls}) {
+    super();
+
+    Hls = externHls;
+
+    this.withCredentials = withCredentials;
+    this.currentTimeFunc = currentTimeFunc;
+    this.mediaSourceMode = mediaSourceMode;
+
+    this.mediaSource = new videojs.MediaSource({ mode: this.mediaSourceMode });
+    // load the media source into the player
+    this.mediaSource.addEventListener('sourceopen', this.handleSourceOpen_.bind(this));
+
+    // combined audio/video or just video when alternate audio track is selected
+    this.mainSegmentLoader_ = new SegmentLoader({
+      mediaSource: this.mediaSource,
+      currentTime: this.currentTimeFunc,
+      withCredentials: this.withCredentials
+    });
+    // alternate audio track
+    this.audioSegmentLoader_ = new SegmentLoader({
+      mediaSource: this.mediaSource,
+      currentTime: this.currentTimeFunc,
+      withCredentials: this.withCredentials
+    });
+
+    this.hlsHandler = hlsHandler;
+    this.hlsHandler.mediaSource = this.mediaSource;
+    this.hlsHandler.selectPlaylist = this.hlsHandler.selectPlaylist || selectPlaylist;
+
+    if (!url) {
+      throw new Error('A non-empty playlist URL is required');
+    }
+
+    this.masterPlaylistLoader_ = new PlaylistLoader(url, this.withCredentials);
+    this.hlsHandler.playlists = this.masterPlaylistLoader_;
+
+    this.masterPlaylistLoader_.on('loadedmetadata', () => {
+      let media = this.masterPlaylistLoader_.media();
+      let master = this.masterPlaylistLoader_.master;
+
+      // if this isn't a live video and preload permits, start
+      // downloading segments
+      if (media.endList &&
+          this.hlsHandler.tech_.preload() !== 'metadata' &&
+          this.hlsHandler.tech_.preload() !== 'none') {
+        this.mainSegmentLoader_.playlist(media);
+        this.mainSegmentLoader_.load();
+      }
+
+      this.audioPlaylistLoaders_ = [];
+      if (master.mediaGroups && master.mediaGroups.AUDIO) {
+        for (let groupKey in master.mediaGroups.AUDIO) {
+          for (let labelKey in master.mediaGroups.AUDIO[groupKey]) {
+            this.audioPlaylistLoaders_.push(new PlaylistLoader(
+              master.mediaGroups.AUDIO[groupKey][labelKey].uri,
+              this.withCredentials,
+              true));
+          }
+        }
+      }
+      this.excludeIncompatibleVariants_(media);
+      this.setupFirstPlay();
+      this.hlsHandler.tech_.trigger('loadedmetadata');
+    });
+
+    this.masterPlaylistLoader_.on('loadedplaylist', () => {
+      let updatedPlaylist = this.masterPlaylistLoader_.media();
+      let seekable;
+
+      if (!updatedPlaylist) {
+        // select the initial variant
+        this.masterPlaylistLoader_.media(this.hlsHandler.selectPlaylist());
+        return;
+      }
+
+      this.mainSegmentLoader_.playlist(updatedPlaylist);
+      this.updateDuration(this.masterPlaylistLoader_.media());
+
+      // update seekable
+      seekable = this.seekable();
```

Comment:
seekable() is a media-element thing, so it seems like it should be calculated by hls.js